### PR TITLE
add http listener

### PR DESCRIPTION
**Pull request by kungyao**

### CODE REVIEW

File: src/HttpListener.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#define HTTPLIB_SRC_HTTPLISTENER_H
+
+#include <string>
+#include <boost/asio.hpp>
+
+/** @brief Create server socket for listening incoming http request. */
+class HttpListener {
+private:
+    /** @brief Required for I/O. */
+    boost::asio::io_context &ioc_;
```
Comment (john0312) [inline src/HttpListener.h:15]: Consider using a pointer.

File: src/HttpListener.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     *  @param error Boost error code.
+     *  @param msg Custom error message.
+     */
+    void handleFail(const boost::system::error_code& error, const std::string& msg);
+    /** @brief Handle accept state.
+     *  @param error Boost error code.
+     */
+    void handleAccept(const boost::system::error_code& error);
+    /** @brief Create HttpSession class object. */
+    void makeHttpSession();
```
Comment (john0312) [inline src/HttpListener.h:30]: Add in comment that you're accepting a socket and creating an http session from it.

File: src/HttpListener.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+public:
+    HttpListener(boost::asio::io_context& ioc, const std::string& ip, const int& port);
+    HttpListener(const HttpListener&) = delete;
+    /** @brief Wait for accept asynchronously. */
+    void runAsyn();
```
Comment (john0312) [inline src/HttpListener.h:35]: Async

File: src/HttpListener.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2021, Kung-Yao Hsieh, Shang-Chen LIN, Yi-Hang Xie. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef HTTPLIB_SRC_HTTPLISTENER_H
+#define HTTPLIB_SRC_HTTPLISTENER_H
+
+#include <string>
+#include <boost/asio.hpp>
```
Comment (john0312) [inline src/HttpListener.h:9]: Please order the include.

File: src/HttpListener.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    /** @brief Create HttpSession class object. 
+     *  
+     *  After acceptor_ accept a new connection, it will create HttpSession for 
+     *  handling request and response by the use of accept_socket_.
+     *  
+     */
+    void makeHttpSession();
+public:
+    HttpListener(boost::asio::io_context& ioc, const std::string& ip, const int& port);
+    HttpListener(const HttpListener&) = delete;
```
Comment (john0312) [inline src/HttpListener.h:38]: Please comment that it's not copyable and delete the copy assignment operator as well.

File: src/HttpListener.h (hunk 4 above)
Comment (kungyao) [reply src/HttpListener.h:10]: done

File: src/HttpListener.h (hunk 5 above)
Comment (kungyao) [reply src/HttpListener.h:40]: done

File: src/HttpListener.h (hunk 3 above)
Comment (kungyao) [reply src/HttpListener.h]: done

File: src/HttpListener.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    void handleAccept(const boost::system::error_code& error);
+    /** @brief Create HttpSession class object. 
+     *  
+     *  After acceptor_ accept a new connection, it will create HttpSession for 
+     *  handling request and response by the use of accept_socket_.
+     *  
+     */
+    void makeHttpSession();
+public:
+    HttpListener(boost::asio::io_context& ioc, const std::string& ip, const int& port);
```
Comment (john0312) [inline src/HttpListener.h:38]: Please use const ref for input or pointer for output.

**Review by john0312 — APPROVED** [review]

**Pull request merged**